server/src/main/java/org/elasticsearch/common/ReferenceDocs.java

```diff
@@ -108,11 +110,31 @@ static Map<String, String> readLinksBySymbol(InputStream inputStream) throws Exc
 
     /**
      * Compute the version component of the URL path (e.g. {@code 8.5} or {@code master}) for a particular version of Elasticsearch. Exposed
-     * for testing, but all items use {@link #VERSION_COMPONENT} ({@code getVersionComponent(Version.CURRENT, Build.CURRENT.isSnapshot())})
-     * which relates to the current version and build.
+     * for testing, but all items use {@link #VERSION_COMPONENT}
+     * ({@code getVersionComponent(Build.current().version(), Build.current().isSnapshot())}) which relates to the current version and
+     * build.
      */
-    static String getVersionComponent(Version version, boolean isSnapshot) {
-        return isSnapshot && version.revision == 0 ? UNRELEASED_VERSION_COMPONENT : version.major + "." + version.minor;
+    static String getVersionComponent(String version, boolean isSnapshot) {
+
+        final Pattern semanticVersionPattern = Pattern.compile("(\\d+)\\.(\\d+)\\.(\\d)+");
+
+        var matcher = semanticVersionPattern.matcher(version);
+        if (matcher.matches()) {
+            var major = matcher.group(1);
+            var minor = matcher.group(2);
+            var revision = matcher.group(3);
+
+            // x.y.z versions with z>0 mean that x.y.0 is released so have a properly versioned docs link
+            if (isSnapshot && "0".equals(revision)) {
+                return UNRELEASED_VERSION_COMPONENT;
+            }
+            return major + "." + minor;
+        }
+        // Non-semantic version
+        if (isSnapshot) {
+            return UNRELEASED_VERSION_COMPONENT;
+        }
+        return CURRENT_VERSION_COMPONENT;
```

Comment:
What is responsible for turning `current` into a sensible url?

Reply:
I don't think I understand your question fully. `current` points at the latest doc version, e.g https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-plugins.html 
I think is a good fallback for when we have no semantic version (e.g. serverless) - it makes sense to point to the latest version.

Reply:
Do you think it makes more sense to always point to `master` instead?

Reply:
Ah right, `current` is valid in itself

Reply:
Might be worth a comment indicating this is the expected scenario for hitting this, and why `current` is the correct place to go